File: example-codes/parking-lot/parking_lot.py
import threading
from .constants import *


class ParkingLot:
    # singleton ParkingLot to ensure only one object of ParkingLot in the system,
    # all entrance panels will use this object to create new parking ticket: get_new_parking_ticket(),
    # similarly exit panels will also use this object to close parking tickets
    instance = None

    class __OnlyOne:
        def __init__(self, name, address):
        # 1. initialize variables: read name, address and parking_rate from database
        # 2. initialize parking floors: read the parking floor map from database,
        #    this map should tell how many parking spots are there on each floor. This
        #    should also initialize max spot counts too.
        # 3. initialize parking spot counts by reading all active tickets from database
        # 4. initialize entrance and exit panels: read from database

            self.__name = name
            self.__address = address
            self.__parking_rate = ParkingRate()

            self.__compact_spot_count = 0
            self.__large_spot_count = 0
            self.__motorbike_spot_count = 0
            self.__electric_spot_count = 0
            self.__max_compact_count = 0
            self.__max_large_count = 0
            self.__max_motorbike_count = 0
            self.__max_electric_count = 0

            self.__entrance_panels = {}
            self.__exit_panels = {}
            self.__parking_floors = {}

            # all active parking tickets, identified by their ticket_number
            self.__active_tickets = {}

            self.__lock = threading.Lock()

    def __init__(self, name, address):
        if not ParkingLot.instance:
            ParkingLot.instance = ParkingLot.__OnlyOne(name, address)
        else:
            ParkingLot.instance.__name = name
            ParkingLot.instance.__address = address

    def __getattr__(self, name):
        return getattr(self.instance, name)

    def get_new_parking_ticket(self, vehicle):
        if self.is_full(vehicle.get_type()):
            raise Exception('Parking full!')
    # synchronizing to allow multiple entrances panels to issue a new
    # parking ticket without interfering with each other
        self.__lock.acquire()
        ticket = ParkingTicket()
        vehicle.assign_ticket(ticket)
        ticket.save_in_DB()
        # if the ticket is successfully saved in the database, we can increment the parking spot count
        self.__increment_spot_count(vehicle.get_type())
        self.__active_tickets.put(ticket.get_ticket_number(), ticket)
        self.__lock.release()
        return ticket

    def is_full(self, type):
        # trucks and vans can only be parked in LargeSpot
        if type == VehicleType.Truck or type == VehicleType.Van:
            return self.__large_spot_count >= self.__max_large_count

        # motorbikes can only be parked at motorbike spots
        if type == VehicleType.Motorbike:
            return self.__motorbike_spot_count >= self.__max_motorbike_count

        # cars can be parked at compact or large spots
        if type == VehicleType.Car:
            return (self.__compact_spot_count + self.__large_spot_count) >= (self.__max_compact_count + self.__max_large_count)

        # electric car can be parked at compact, large or electric spots
        return (self.__compact_spot_count + self.__large_spot_count + self.__electric_spot_count) >= (self.__max_compact_count + self.__max_large_count
                                                                                                  + self.__max_electric_count)

    # increment the parking spot count based on the vehicle type
    def increment_spot_count(self, type):
        large_spot_count = 0
        motorbike_spot_count = 0
        compact_spot_count = 0
        electric_spot_count = 0
        if type == VehicleType.Truck or type == VehicleType.Van:
            large_spot_count += 1
        elif type == VehicleType.Motorbike:
            motorbike_spot_count += 1
        elif type == VehicleType.Car:
            if self.__compact_spot_count < self.__max_compact_count:
                compact_spot_count += 1
            else:
                large_spot_count += 1
        else:  # electric car
            if self.__electric_spot_count < self.__max_electric_count:
                electric_spot_count += 1
            elif self.__compact_spot_count < self.__max_compact_count:
                compact_spot_count += 1
            else:
                large_spot_count += 1

        def is_full(self):
            for key in self.__parking_floors:
                if not self.__parking_floors.get(key).is_full():
                    return False
            return True

        def add_parking_floor(self, floor):
            # store in database
            None

        def add_entrance_panel(self, entrance_panel):
            # store in database
            None

        def add_exit_panel(self,  exit_panel):
            # store in database
            None

File: example-codes/parking-lot/java/ParkingLot.java
package parkinglot;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Singleton class representing the parking lot system.
 * Thread-safe implementation using double-checked locking.
 */
public class ParkingLot {
    private static ParkingLot instance = null;
    private static final Lock lock = new ReentrantLock();

    private String name;
    private String address;
    private ParkingRate parkingRate;

    private int compactSpotCount;
    private int largeSpotCount;
    private int motorbikeSpotCount;
    private int electricSpotCount;
    private int maxCompactCount;
    private int maxLargeCount;
    private int maxMotorbikeCount;
    private int maxElectricCount;

    private Map<String, EntrancePanel> entrancePanels;
    private Map<String, ExitPanel> exitPanels;
    private Map<String, ParkingFloor> parkingFloors;
    private Map<String, ParkingTicket> activeTickets;

    private ParkingLot(String name, String address) {
        this.name = name;
        this.address = address;
        this.parkingRate = new ParkingRate();
        this.entrancePanels = new HashMap<>();
        this.exitPanels = new HashMap<>();
        this.parkingFloors = new HashMap<>();
        this.activeTickets = new HashMap<>();
    }

    /**
     * Thread-safe singleton instance getter.
     */
    public static ParkingLot getInstance(String name, String address) {
        if (instance == null) {
            lock.lock();
            try {
                if (instance == null) {
                    instance = new ParkingLot(name, address);
                }
            } finally {
                lock.unlock();
            }
        }
        return instance;
    }

    public boolean isFull(ParkingSpotType type) {
        if (type == ParkingSpotType.COMPACT) {
            return compactSpotCount >= maxCompactCount;
        } else if (type == ParkingSpotType.LARGE) {
            return largeSpotCount >= maxLargeCount;
        } else if (type == ParkingSpotType.MOTORBIKE) {
            return motorbikeSpotCount >= maxMotorbikeCount;
        } else if (type == ParkingSpotType.ELECTRIC) {
            return electricSpotCount >= maxElectricCount;
        }
        return true;
    }

    public ParkingTicket getNewParkingTicket(Vehicle vehicle) {
        if (this.isFull(getParkingSpotTypeForVehicle(vehicle.getType()))) {
            System.out.println("Parking full!");
            return null;
        }

        lock.lock();
        try {
            ParkingTicket ticket = new ParkingTicket();
            vehicle.assignTicket(ticket);
            ticket.setTicketNumber(generateTicketNumber());
            activeTickets.put(ticket.getTicketNumber(), ticket);
            return ticket;
        } finally {
            lock.unlock();
        }
    }

    public boolean isFull() {
        return compactSpotCount >= maxCompactCount &&
               largeSpotCount >= maxLargeCount &&
               motorbikeSpotCount >= maxMotorbikeCount &&
               electricSpotCount >= maxElectricCount;
    }

    private ParkingSpotType getParkingSpotTypeForVehicle(VehicleType type) {
        switch (type) {
            case CAR:
                return ParkingSpotType.COMPACT;
            case TRUCK:
            case VAN:
                return ParkingSpotType.LARGE;
            case MOTORBIKE:
                return ParkingSpotType.MOTORBIKE;
            case ELECTRIC:
                return ParkingSpotType.ELECTRIC;
            default:
                return ParkingSpotType.COMPACT;
        }
    }

    private String generateTicketNumber() {
        return "TICKET-" + System.currentTimeMillis();
    }
}

class ParkingRate {
    // Parking rate calculation logic
}

class EntrancePanel {
    // Entrance panel logic
}

class ExitPanel {
    // Exit panel logic
}
